backend/data/users.js

```diff
@@ -270,35 +270,92 @@ async function getCourse(id) {
+ * @param {string} assignmentId 
+ * @param {string} grade 
+ * @returns object
+ */
 async function addGrade(studentId, courseId, assignmentId, grade) {
-  error.str(studentId);
   const parsedStudentId = error.validId(studentId);
-  error.str(assignmentId);
   const parsedAssignmentId = error.validId(assignmentId);
-  error.str(grade);
+  const parsedCourseId = error.validId(courseId);
   const parsedGrade = parseInt(grade);
 
   const userCollection = await users();
   const user = await getUser(studentId);
+  if (!user) throw new Error('No user was found with the given id');
   const course = await getCourse(courseId);
+  if (!course) throw new Error('No course was found with the given id');
 
-  let classObj = user.classes.find((e) => e._id === courseId);
+  let classObj = user.classes.find((e) => e._id.toString() === courseId);
   if (!classObj) throw new Error('Student is not enrolled in the course');
   if (!course.assignments.some((e) => e._id.toString() === assignmentId)) throw new Error('No assignment found');
-  if (!classObj.grades.some((e) => e._id === assignmentId)) throw new Error('No assignment found for the user');
+  if (!classObj.grades.some((e) => e._id.toString() === assignmentId)) throw new Error('No assignment found for the user');
   if (grade < 0 || grade > 100) throw new Error('Grade should be betwenn 0 - 100');
 
-  const query = {
-    $and: [{ _id: parsedStudentId }, { 'grades._id': parsedAssignmentId }],
-  };
-
-  //"push" the updated course info to the same ID in the database
-  const updatedInfo = await userCollection.updateOne(query, { $set: { grade: parsedGrade } });
+  const updatedInfo = await userCollection.findOneAndUpdate(
+    { _id: parsedStudentId },
+    { $set: { "classes.$[classId].grades.$[assign].grade": parsedGrade } },
+    { arrayFilters: [{ 'classId._id': parsedCourseId }, { 'assign._id': parsedAssignmentId }] }
+  );
 
   //check that the update succeeded
   if (updatedInfo.modifiedCount === 0) throw new Error('Failed to add grade for assignment');
+  const overallGradeset = await setOverallGrade(studentId, courseId);
+  if (!overallGradeset) throw new Error('Unable to calculate overall grade');
+  return { "gradeAdded": true };
+}
+
+/**
+ * Calculate overall grade
+ * @param {string} studentId 
+ * @param {string} courseId 
+ * @returns object
+ */
+async function setOverallGrade(studentId, courseId) {
+  error.str(studentId);
+  const parsedStudentId = error.validId(studentId);
+  error.str(courseId);
+  const parsedCourseId = error.validId(courseId);
+
+  const userCollection = await users();
+  const user = await getUser(studentId);
+  if (!user) throw new Error('No user was found with the given id');
+  const course = await getCourse(courseId);
+  if (!course) throw new Error('No course was found with the given id');
+
+  let classObj = user.classes.find((e) => e._id.toString() === courseId);
+  if (!classObj) throw new Error('Student is not enrolled in the course');
+
```

Comment:
Shouldn't be necessary if we do everything else right, but would probably be smart to throw in a check to see if there exists at least 1 "grade" entry for the class just to make sure in case there somehow is, that we don't end up dividing by 0 in the following calculation.

Reply:
`if (count > 0) overallGrade = Math.round(sum / count * 1e2) / 1e2;`
I have added the above if condition to handle scenarios where there are no grades